JavaScript Providers on Java 17 (#1537)
keycloak/keycloak-documentation: server_development/topics/providers.adoc

Once you have a JAR file with a descriptor and the scripts you want to deploy, you just need to copy the JAR to the {project_name} `standalone/deployments/` directory.
endif::[]

===== Deploy the script engine on Java 15 and later

> In order to run the scripts, JavaScript providers require JavaScript engine to be available in your Java application. In the Java 14 or earlier, there is Nashorn JavaScript Engine
> automatically available as part of the Java itself and JavaScript providers are able to leverage this script engine out of the box. However in case of Java 15 or later, script engine is not part
> of the Java itself and needs to be added to your server. {project_name} does not have any script engine by default. So in case of Java 15 or later, there is additional step needed when deploying script

Reviewer: ```suggestion
To run the scripts, JavaScript providers require that a JavaScript engine is available in your Java application. Java 14 and lower versions include the Nashorn JavaScript Engine. It is 
automatically available as part of the Java itself and JavaScript providers are able to use this script engine by default. However, for Java 15 or higher versions, the script engine is not part
of the Java itself. It needs to be added to your server because {project_name} does not have any script engine by default. Java 15 and higher versions require an extra step when deploying script
```

providers - adding the script engine of your choice to your distribution.

> You can use any script engine you want, however we test just with the Nashorn JavaScript Engine and in the steps below, it is assumed that this one will be used:

Reviewer: ```suggestion
You can use any script engine. However, we only test with the Nashorn JavaScript Engine. The following steps assume that this engine is used:
```

ifeval::["{kc_dist}" == "quarkus"]

> You can install script engine by copying the nashorn script engine JAR together with it's dependencies directly to the `KEYCLOAK_HOME/providers` directory. In the `pom.xml` file

Reviewer: ```suggestion
Install the script engine by copying the nashorn script engine JAR and its dependencies directly to the `KEYCLOAK_HOME/providers` directory. In the `pom.xml` file
```

> of your script project, you can declare the dependency like this in the `dependencies` section:

Reviewer: ```suggestion
of your script project, you can declare the dependency such as this in the `dependencies` section:
```

```xml
<dependency>
    <groupId>org.openjdk.nashorn</groupId>
    <artifactId>nashorn-core</artifactId>
    <version>15.3</version>
</dependency>
```

and declare `maven-dependency-plugin` in the `plugins` section to copy the dependencies to the specified directory:

```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-dependency-plugin</artifactId>
    <executions>
        <execution>
                <id>copy-dependencies-quarkus</id>
                <phase>package</phase>
            <goals>
                <goal>copy-dependencies</goal>
            </goals>
            <configuration>
                <outputDirectory>${project.build.directory}/keycloak-server-copy/providers</outputDirectory>
                <includeArtifactIds>nashorn-core,asm,asm-util,asm-commons</includeArtifactIds>
            </configuration>
        </execution>
    </executions>
</plugin>
```

> Once the project is built, you can copy the script engine together with it's dependencies to the `KEYCLOAK_HOME/providers` directory.

Reviewer: ```suggestion
Once the project is built, copy the script engine and its dependencies to the `KEYCLOAK_HOME/providers` directory.
```

```bash
cp target/keycloak-server-copy/providers/*.jar KEYCLOAK_HOME/providers/
```

> After re-augment the distribution with `kc.sh build` the script engine should be deployed and your script providers should work.

Reviewer: ```suggestion
After re-augment, the distribution with `kc.sh build` the script engine should be deployed and your script providers should work.
```

Author: I am not sure about this one change. Let me explain: re-augmenting of the distribution itself is done by running `kc.sh build` script. So it is same like if I say for example:
`After changing the house with new bricks` (Probably dummy example, but I don't have any better in my mind :-) ).

So I wonder if the comma should be before the second `the` ? For example:
`After re-augment the distribution with `kc.sh build`, the script engine should be deployed and your script providers should work.`

WDYT?

Reviewer: Exactly.  I was just looking at this and coming to the same conclusion.  Thanks for correcting this, @mposolda.

endif::[]

ifeval::["{kc_dist}" == "wildfly"]

> You can install script engine by adding new module nashorn-core to your {project_name}. When server is up and running, you can run commands similar to these in the `KEYCLOAK_HOME/bin` directory:

Reviewer: ```suggestion
You can install the script engine by adding the new module nashorn-core to your {project_name}. After the server starts, you can run commands similar to these in the `KEYCLOAK_HOME/bin` directory:
```

```bash
export NASHORN_VERSION=15.3
wget https://repo1.maven.org/maven2/org/openjdk/nashorn/nashorn-core/$NASHORN_VERSION/nashorn-core-$NASHORN_VERSION.jar
./jboss-cli.sh -c --command="module add --module-root-dir=../modules/system/layers/keycloak/ --name=org.openjdk.nashorn.nashorn-core --resources=./nashorn-core-$NASHORN_VERSION.jar --dependencies=asm.asm,jdk.dynalink"
rm nashorn-core-$NASHORN_VERSION.jar
```

In case you want to install your provider into the different module, you can use configuration property `script-engine-module` of the default scripting provider.

> For example you can use something like this in your `KEYCLOAK_HOME/standalone/configuration/standalone-*.xml` file:

Reviewer: ```suggestion
For example, you can use something such as this in your `KEYCLOAK_HOME/standalone/configuration/standalone-*.xml` file:
```
